hubble/cmd/common/conn/conn.go

```diff
@@ -30,13 +34,100 @@ var GRPCOptionFuncs []GRPCOptionFunc
 func init() {
 	GRPCOptionFuncs = append(
 		GRPCOptionFuncs,
-		grpcInterceptors,
+		grpcUnaryInterceptors,
+		grpcStreamInterceptors,
 		grpcOptionTLS,
 	)
 }
 
-func grpcInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
-	return grpc.WithUnaryInterceptor(timeout.UnaryClientInterceptor(vp.GetDuration(config.KeyRequestTimeout))), nil
+func grpcUnaryInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
+	option := grpc.WithChainUnaryInterceptor(
+		timeout.UnaryClientInterceptor(vp.GetDuration(config.KeyRequestTimeout)),
+		logVersionMismatchUnaryInterceptor(),
+	)
+	return option, nil
+}
+
+func grpcStreamInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
+	option := grpc.WithChainStreamInterceptor(
+		logVersionMismatchStreamInterceptor(),
+	)
+	return option, nil
+}
+
+func logVersionMismatchUnaryInterceptor() grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req any, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		var header metadata.MD
+		opts = append(opts, grpc.Header(&header))
+		if err := invoker(ctx, method, req, reply, cc, opts...); err != nil {
+			return err
+		}
+		logVersionMismatch(header)
+		return nil
+	}
+}
+
+func logVersionMismatchStreamInterceptor() grpc.StreamClientInterceptor {
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		stream, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, err
+		}
+		// stream.Header() blocks until metadata is ready to read.
+		// This could be forever if no metadata is ever sent, after
+		// the first read from the stream or after an explicit call
+		// to SendHeader() server-side.
+		go func() {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			header, err := stream.Header()
+			if err == nil {
+				logVersionMismatch(header)
+			}
+		}()
+		return stream, nil
+	}
+}
+
+func logVersionMismatch(header metadata.MD) {
```

Comment:
It gets the job done, but I'll take suggestion to improve or react differently in here :)

Reply:
I think it's fine as is, just bike-shedding here:

IMO, we could just unconditionally debug log the relay and server version. It's a debug log, so I think that would be fine.

If we want to keep the current logging on mismatch, I'd probably only log if the minor version differs. Patch version mismatch don't really make a difference and the current log line makes it feel like that's an issue, which it really should be

Reply:
As suggested by you and @kaworu, I've updated to only log on minor version mismatch at minimum.